docs/arch/02-core-concepts.md

```diff
@@ -0,0 +1,693 @@
+
+**Related concepts:** Stdio Transport, Runtime
+
+### Parse
+
+**Parse** extracts structured information from JSON-RPC MCP messages for middleware processing.
+
+**Parsing includes:**
+- Message type (request, response, notification)
+- Method name (e.g., `tools/call`, `resources/read`)
+- Request ID
+- Parameters
+- Resource ID (for resource operations)
+- Arguments (for tool calls)
+
+**Parsed data stored in context:**
+- Available to downstream middleware
+- Used by authorization for policy evaluation
+- Used by audit for event logging
+
+**Implementation:**
+- Parser middleware: `pkg/mcp/parser.go`
+- Parsed type: `pkg/mcp/types.go`
+
+**Related concepts:** Middleware, Authorization, Audit
+
+### Filter and Override
+
+**Filter and override** controls which tools are available to MCP clients and how they are presented.
+
+**Two complementary operations:**
+
+1. **Tool Filtering**: Whitelist specific tools by name
+   - Configured via `--tool` flags or `toolsFilter` config
+   - Tools not in filter list are hidden from clients
+   - Empty filter list means all tools are available
+
+2. **Tool Overriding**: Customize tool presentation
+   - Configured via `toolsOverride` map in config file
+   - Override tool names and/or descriptions
+   - Maps actual tool name to user-visible name/description
+
+**Two middlewares for consistency:**
+
+- **Tool Filter middleware**: Processes outgoing `tools/list` responses
+- **Tool Call Filter middleware**: Processes incoming `tools/call` requests
+
+Both middlewares share the same configuration to ensure clients only see tools they can call, and can only call tools they see.
+
+**Configuration:**
+- `toolsFilter` - List of allowed tool names (from `--tool` flags)
+- `toolsOverride` - Map from actual name to override (from config file)
+
+**Implementation:**
+- Middleware factories: `pkg/mcp/middleware.go`
+- Filter logic: `pkg/mcp/tool_filter.go`
+- Configuration: `pkg/runner/config.go`
+
+**Related concepts:** Middleware, Authorization
+
+### Authorize
+
+**Authorize** evaluates Cedar policies to determine if requests are permitted.
+
+**Authorization process:**
+1. Get parsed MCP data from context
+2. Get JWT claims from auth middleware
+3. Create Cedar entities (Principal, Action, Resource)
+4. Evaluate Cedar policies
+5. Allow or deny request
+
+**Policy language:**
+Cedar policies use:
+- `principal` - Who is making the request (from JWT)
+- `action` - What operation (from MCP method)
+- `resource` - What is being accessed (from MCP resource ID)
+
+**Example policy:**
+```cedar
+permit(
+  principal == Client::"user@example.com",
+  action == Action::call_tool,
+  resource == Tool::"web-search"
+);
+```
+
+**Implementation:**
+- Authz middleware: `pkg/authz/middleware.go`
+- Policy engine: Cedar (external library)
+
+**Related concepts:** Middleware, Authentication, Parse
+
+### Audit
+
+**Audit** logs MCP operations for compliance, monitoring, and debugging.
+
+**Audit events include:**
```

Comment:
The audit event type list is incomplete. The documentation lists 8-9 examples but states there are 15 total event types.

Missing event types: `mcp_notification`, `mcp_ping`, `mcp_logging`, `mcp_completion`, `mcp_roots_list_changed`, `mcp_request`.

Consider either listing all 15 event types explicitly or providing a more representative sample.

Evidence: pkg/audit/mcp_events.go